components/ws-daemon/pkg/iws/iws.go

```diff
@@ -995,6 +1002,294 @@ func (wbs *InWorkspaceServiceServer) unPrepareForUserNS() error {
+		if err != nil {
+			return nil, err
+		}
+
+		memory, err := getMemoryResourceInfoV1(mountPoint, cgroupPath)
+		if err != nil {
+			return nil, err
+		}
+
+		return &api.Resources{
+			Cpu:    cpu,
+			Memory: memory,
+		}, nil
+	}
+}
+
+func getCpuResourceInfoV2(mountPoint, cgroupPath string) (*api.Cpu, error) {
+	cpu := v2.NewCpuControllerWithMount(mountPoint, cgroupPath)
+
+	t, err := resolveCPUStatV2(cpu)
+	if err != nil {
+		return nil, err
+	}
+
+	time.Sleep(time.Second)
+
+	t2, err := resolveCPUStatV2(cpu)
+	if err != nil {
+		return nil, err
+	}
+
+	cpuUsage := t2.usage - t.usage
+	totalTime := t2.uptime - t.uptime
+	used := cpuUsage / totalTime * 1000
+
+	quota, period, err := cpu.Max()
+	if err != nil {
+		return nil, err
+	}
+
+	// if no cpu limit has been specified, use the number of cores
+	var limit uint64
+	if quota == math.MaxUint64 {
+		cpuInfo, err := linuxproc.ReadCPUInfo("/proc/cpuinfo")
+		if err != nil {
+			return nil, err
+		}
+
+		limit = uint64(cpuInfo.NumCore()) * 1000
+	} else {
+		limit = quota / period * 1000
+	}
+
+	return &api.Cpu{
+		Used:  int64(used),
+		Limit: int64(limit),
+	}, nil
+}
+
+func getMemoryResourceInfoV2(mountPoint, cgroupPath string) (*api.Memory, error) {
+	memory := v2.NewMemoryControllerWithMount(mountPoint, cgroupPath)
+	memoryLimit, err := memory.Max()
+	if err != nil {
+		return nil, xerrors.Errorf("could not retrieve memory max: %w", err)
+	}
+
+	memInfo, err := linuxproc.ReadMemInfo("/proc/meminfo")
+	if err != nil {
+		return nil, xerrors.Errorf("failed to read meminfo: %w", err)
+	}
+
+	// if no memory limit has been specified, use total available memory
+	if memoryLimit == math.MaxUint64 || memoryLimit > memInfo.MemTotal*1024 {
+		// total memory is specifed on kilobytes -> convert to bytes
+		memoryLimit = memInfo.MemTotal * 1024
+	}
+
+	usedMemory, err := memory.Current()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to read current memory usage: %w", err)
+	}
+
+	return &api.Memory{
+		Limit: int64(memoryLimit),
+		Used:  int64(usedMemory),
+	}, nil
+}
+
+func getMemoryResourceInfoV1(mountPoint, cgroupPath string) (*api.Memory, error) {
+	memory := v1.NewMemoryControllerWithMount(mountPoint, cgroupPath)
+
+	memoryLimit, err := memory.Limit()
+	if err != nil {
+		return nil, err
+	}
+
+	memInfo, err := linuxproc.ReadMemInfo("/proc/meminfo")
+	if err != nil {
+		return nil, xerrors.Errorf("failed to read meminfo: %w", err)
+	}
+
+	// if no memory limit has been specified, use total available memory
+	if memoryLimit == math.MaxUint64 || memoryLimit > memInfo.MemTotal*1024 {
+		// total memory is specifed on kilobytes -> convert to bytes
+		memoryLimit = memInfo.MemTotal * 1024
+	}
+
+	usedMemory, err := memory.Usage()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to read memory limit: %w", err)
+	}
+
+	stats, err := memory.Stat()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to read memory stats: %w", err)
+	}
+
+	if stats.InactiveFileTotal > 0 {
```

Comment:
it is not applicable to cgroup v2?

An idea is to inform users when they are really close to OOM kills. Inactive files should not be counted in used, since they can be released. Is it different for v2?

Reply:
It makes sense to bring it in line with cgroup v1. Looking through the documentation for memory.current it includes page cache, so we should subtract it if the goal is to show how close they are to OOM.

Reply:
fyi I originally learned it from here: https://github.com/kubernetes/kubernetes/blob/d5e5f14508b02997b117ca9214b2fa349e57d0f9/test/e2e/node/node_problem_detector.go#L259-L293

I was investigating how GCP is showing, since they subtract evictable memory. And following the logic node problem detector I was able to get the same number like GCP: